### ISSUE TITLE
hlint v3.0.4: Parse error for `NoStarIsType` extension

### ISSUE THREAD

**Issue by jinwoo** [ndmitchell/hlint#971]
When hlint v3.0.4 parses an import statement like below, it gives a parse error:

```
{-# LANGUAGE NoStarIsType #-}
...

import GHC.TypeLits(KnownNat, type (+), type (*))
...
```

```
$ hlint Foo.hs 
Foo.hs:126:47: Error: Parse error: on input `*'
Found:
    import GHC.Generics(Generic, Generic1)
    import GHC.Stack
  > import GHC.TypeLits(KnownNat, type (+), type (*))
    import qualified GHC.TypeLits as T
    import Test.QuickCheck
  

1 hint
```

Explicitly passing `-XNoStarIsType` to hlint doesn't help either.

It used to work fine with v2.2.11.

**Comment by shayne-fletcher** [comment]
I'm a little confused by this.

In so far as I know, one can't disable a language pragma at source (happy to be told I'm wrong) so,
```haskell
{-# LANGUAGE NoStarIsType #-}
```
Looks wrong to me. 

Now suppose this program,
```haskell
import GHC.TypeLits(KnownNat, type (+), type (*))

main = undefined
```
Then compiling with `ghc -c -XNoStarIsType`, I get
```
/Users/shaynefletcher/Quux.hs:2:36: error:
    Illegal keyword 'type' (use ExplicitNamespaces to enable)
  |
2 | import GHC.TypeLits(KnownNat, type (+), type (*))
  |                                    ^^^

/Users/shaynefletcher/Quux.hs:2:46: error:
    Illegal keyword 'type' (use ExplicitNamespaces to enable)
  |
2 | import GHC.TypeLits(KnownNat, type (+), type (*))
  |                                              ^^^
```
If I modify the program to read,
```haskell
{-# LANGUAGE ExplicitNamespaces #-}

import GHC.TypeLits(KnownNat, type (+), type (*))

main = undefined
```
and compile with `ghc -XStarIsType -c ~/Quux.hs`, I get
```
/Users/shaynefletcher/Quux.hs:3:47: error: parse error on input ‘*’
  |
3 | import GHC.TypeLits(KnownNat, type (+), type (*))
  |           
```
and with `ghc -c ~/Quux.hs` (or equivalently) `ghc -c -XNoStarIsType ~/Quux.hs`
all is well (that is, I get a successful parse).

In summary, right now it appears to me that this is not an HLint issue, rather we are looking at GHC compilation failures that happen to be reported by HLint (since it's now using `ghc-lib` to 
 rather than HSE which was presumably accepting invalid parses).

**Comment by shayne-fletcher** [comment]
Actually, this issue has highlighted a bug in `ghc-lib-parser-ex`. Regardless of the fact that  HLint is default enabling `StarIsType`, the presence of `{-# LANGUAGE NoStarIsType #-}` should have the effect of disabling it.

**Comment by shayne-fletcher** [comment]
PR https://github.com/ndmitchell/hlint/pull/973 resolves the `ghc-lib-parser-ex` issue.

**Comment by jinwoo** [comment]
Thanks for the quick fix, @shayne-fletcher !

Sorry I didn't realize that `ExplicitNamespaces` is also needed. Our actual code uses `TypeFamilies` and `TypeOperators`, which imply `ExplicitNamespaces`.

Re: default enabling/disabling `StarIsType`, `StarIsType` is enabled by default (for a historical reason) but can be disabled `NoStarIsType` AFAIK. https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-StarIsType.

I'm not sure what the implication is of not enabling `StarIsType` by default though.

**Comment by shayne-fletcher** [comment]
Thanks for pointing me to that info @jinwoo. I've put a not on PR https://github.com/ndmitchell/hlint/pull/972 accordingly indicating that there is some contention over whether that PR should be applied or not.

**Comment by jinwoo** [comment]
Yeah, I think your fix in ghc-lib-parser-ex will be enough to fix this issue: https://github.com/shayne-fletcher/ghc-lib-parser-ex/commit/ff651e133d25eb3abbbe4eb3ae1d0bf57d22ad5e

Thanks for that!

**Comment by ndmitchell** [comment]
Thanks all! Nice work tracking it down and fixing. I've added a test and will make a new release when I fix up 3 other bugs that are on my nearterm radar.

**Comment by jinwoo** [comment]
I retried after importing ghc-lib-parser-ex 8.10.0.6, but I'm still getting the same parse error from hlint. Can this be reopened?

**Comment by jinwoo** [comment]
And @ndmitchell , you need `ExplicitNamespaces` (or `TypeOperators` or `TypeFamilies`) in your test (it's failing due to that).

**Comment by jinwoo** [comment]
I traced this and the parse error comes from the `parseFile` function in https://github.com/shayne-fletcher/ghc-lib-parser-ex/blob/master/src/Language/Haskell/GhclibParserEx/GHC/Parser.hs#L134

It's called from `fileToModule` in https://github.com/ndmitchell/hlint/blob/master/src/GHC/Util.hs#L49, which in turn is called from `parseModuleEx` in https://github.com/ndmitchell/hlint/blob/master/src/HSE/All.hs#L160

**Comment by ndmitchell** [comment]
@jinwoo can you give me a sample file that fails?

**Comment by jinwoo** [comment]
I think I know what the problem is.

When I add `{-# LANGUAGE NoStarIsType #-}` in the file, hlint works fine.

We internally patch hlint so that it uses some extensions by default. We patch the `cmdExtensions` function in https://github.com/ndmitchell/hlint/blob/master/src/CmdLine.hs#L234

I see this problem also when I pass `-XNoStarIsType`.

So it looks like the command-line flags don't have any effects.

**Comment by shayne-fletcher** [comment]
I can't reproduce this HEAD. My test program is,
```haskell
{-# LANGUAGE NoStarIsType #-}
{-# LANGUAGE ExplicitNamespaces #-}

import GHC.TypeLits(KnownNat, type (+), type (*))

main = undefined
```
When I run it through hlint I get,
```
Found:
  {-# LANGUAGE NoStarIsType #-}
Note: may break the code

/Users/shaynefletcher/Quux.hs:2:1-35: Warning: Avoid restricted extensions
Found:
  {-# LANGUAGE ExplicitNamespaces #-}
Note: may break the code

2 hints
```
@jinwoo Are you sure that HLint has been rebuilt against ghc-lib-parser-ex-8.10.0.6?

**Comment by jinwoo** [comment]
@shayne-fletcher As I noted above, now the problem seems to be in hlint, rather than in ghc-parser-lib-ex.

**Comment by shayne-fletcher** [comment]
> When I add `{-# LANGUAGE NoStarIsType #-}` in the file, hlint works fine.

Ok. That's the expected behavior. HLint is going to enable `StarIsType` by default (consistent with GHC defaults) so you continue to need to explicitly disable it via the pragma above.

**Comment by jinwoo** [comment]
But the command-line flag `-XNoStarIsType` should also work, no?

**Comment by shayne-fletcher** [comment]
> But the command-line flag `-XNoStarIsType` should also work, no?

That comes down to the settings/command line configuration and the set of base flags produced by those means. I'm less familiar with that area of the code base and would have to look into that. Right now I'm going to go out on a limb and say apparently not 😄

**Comment by shayne-fletcher** [comment]
A quick look at `getExtensions` and I think it needs some work. Ideally it should be returning a set of extensions to enable and a set of extensions to disable. Right now it appears to only return extensions that to enable.

**Comment by jinwoo** [comment]
Would it work to detect the "No" prefix and consider it as disabling the corresponding extension? I think that's what GHC does: https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#language-options

> Every language option can switched on by a command-line flag “-X...” (e.g. -XTemplateHaskell), and switched off by the flag “-XNo...”; (e.g. -XNoTemplateHaskell).

**Comment by shayne-fletcher** [comment]
Yes, the code to do that has just been poorly adapted from HSE it seems and needs some attention.

**Comment by shayne-fletcher** [comment]
I've quickly hacked a patch. Hold while I test it.

**Comment by shayne-fletcher** [comment]
I have offered a fix in PR https://github.com/ndmitchell/hlint/pull/974. @jinwoo if you have the means, you can test it solves your problem by building hlint from the `harvest-disabled-extensions` branch.

**Comment by jinwoo** [comment]
@shayne-fletcher Thanks. I'll patch in the fix and check.

**Comment by jinwoo** [comment]
The PR looks good! Thanks.

**Comment by jinwoo** [comment]
Does https://github.com/ndmitchell/hlint/pull/975 also fix this issue?

**Comment by jinwoo** [comment]
Answering my own question. Probably not. #975 is only about hints.

**Comment by shayne-fletcher** [comment]
To confirm, no it's unrelated. #975 is more in the spirit of https://github.com/ndmitchell/hlint/commit/c02daf7da1f61d0a0f8abebf06f344caf57463dc than this.

**Comment by ndmitchell** [comment]
I believe  #974 should fix it. The issue was that NoStarInType is a default-on extension. Given HLint sets nearly every extension, we treated -XNoFoo as "don't turn on Foo", but for default-on extensions you have to tell GHC to explicitly turn them off.

**Comment by ndmitchell** [comment]
@jinwoo can you try again now? I appreciate we've failed to solve this more than once now. Also, any chance you could use the `.hlint.yaml` file setting? You can set various extensions on/off there, which might let you avoid patching HLint.

**Comment by jinwoo** [comment]
@ndmitchell Thanks for merging this quickly. As I noted above, I patched in @shayne-fletcher 's PR locally and verified that it fixed this problem.

Re: `.hlint.yaml`, yeah I agree it's better than patching hlint. But our current linting infrastructure is huge (it includes all the languages that we use) and we couldn't come up with a solution yet. We eventually want to support that.

Thanks again, everyone, for the quick fix!

**Comment by ndmitchell** [comment]
Great, thanks for the verification. Yep, I imagine your linting is an amazing beast - glad HLint can be part of it :)
